DietPi-Backup/Sync | Add free space check before run

MichaIng opened the pull request:
+ DietPi-Backup | Add free space check before running backup: https://github.com/Fourdee/DietPi/issues/1664
+ DietPi-Backup | Adjust rsync options: Remove "l" and "p" as included in "a" and "P" as we do not show direct output anymore

This is early **WIP** just to review the technique, do speed tests etc.
- Found `(( ))` not allow calculation with decimal values, nor does bash even support decimal variables? `bc` seems to be a solution for this, but I was a bid tired to learn it and found `* 12 / 10` as an easy replacement for `* 1.2`
- Found `G_CHECK_FREESPACE` and `G_CHECK_VALIDINT` a bid inconsistent currently in use. If `return $return_value` is used, the function cannot be used within arithmetic checks, but direct checks, as used within this PR are possible. But for this use, the output should be switched, as `0` returns as true and anything else (?) as false. `G_CHECK_VALIDINT` uses `echo` instead of `return`, which then can be used in arithmetic checks and `if (( $(G_CHECK_VALIDINT 10) )); then` executes code as expected, but I guess this is the slower method? Will do some benchmarking in the code performance issue 😋.
- [x] Currently rsync --dry-run also logs to logfile, which is not wanted I guess. 
- [x] If system backup is cancelled, userdata backup cancelled silently as well, but the other way round, if system backup is made, userdata backup can be still cancelled independently.
- [x] `G_WHIP_MSG` currently shows `completed`, even if backup was cancelled due to insufficient free space.
- [x] Should be integrated into DietPi-Sync as well, if method passes performance and reliability tests.

Indeed rsync `total size` result is always smaller than the `du` result afterwards, I guess as rsync shows theoretical file size (transferred bytes), while du shows actual used disc space based on e.g. 4k blocks. No idea if there is some better way to estimate this. Theoretically if a huge amount of 1 byte files is transferred, the actual used size will be 4000 times larger than what rsync shows 😆. But `1.2` security factor should work in most real cases at least. Good enough for the effort?

Commit:
v6.7

+ DietPi-Backup | Add free space check before running backup: https://github.com/Fourdee/DietPi/issues/1664
+ DietPi-Backup | Adjust rsync options: Remove "l" and "p" as included in "a" and "P" as we do not show direct output anymore

Fourdee commented:
@MichaIng 

> nor does bash even support decimal variables? bc seems to be a solution for this

Yep, bash does not support floating point calculations, only integer. ```bc``` is needed for floats, or, remove the decimal point and its contents.

Fourdee commented:
@MichaIng 

Looks good, great work 👍 

I'll merge for testing.

Fourdee commented:
Interesting, ```bc``` doesn't handle ```scale=0``` if input contains any floats:
```
root@DietPi:~# echo "scale=0; $new_backup_size * 1.21230" | bc
1495.97820

root@DietPi:~# echo "scale=0; 13 / 9" | bc
1
```

Solution is to ```/1``` at the end to trigger ```bc``` adjust.
```
root@DietPi:~# echo "scale=0; $new_backup_size * 1.21230 / 1" | bc
1495
```

Fourdee commented:
We can use ```du -m``` for MB result, rounds up:
```
root@DietPi:~# du -s /mnt/dietpi-backup/
594940  /mnt/dietpi-backup/
# = 580.99609375 MB

root@DietPi:~# du -ms /mnt/dietpi-backup/
581     /mnt/dietpi-backup/

```

MichaIng commented:
@Fourdee 
- Jep looks good now 😃. Runs great on VM, interesting would be speed tests on slow SBCs with sd card.
- Just did some size estimation check:
```
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_userdata | available=9888 MB | required=183 MB
...
2018/04/09 14:31:06 [2806] total size is 162,402,959  speedup is 0.99
...
root@DietPi:~# du -sm //mnt/dietpi-backup/dietpi-backup_userdata/
196     //mnt/dietpi-backup/dietpi-backup_userdata/
```
- So even with security factor, the final backup size was >7% larger. There was mainly the Nextcloud installation backup inside, with many folders and small PHP files inside. So yeah, prediction without actual transferring is quite difficult and not too reliable, but maybe better than nothing. The alternative from my point of view would be to skip the test, but removing the whole backup (to have free space available, not leaving user with an uncomplete/unusable backup) if an error occurs.

Fourdee commented:
@MichaIng 

> So even with security factor, the final backup size was >7% larger. There was mainly the Nextcloud installation backup inside, with many folders and small PHP files inside

Interesting.

The only thing I can think of is:
- Add ```4K``` for the total number of files listed during dry run. However, cant see a total file count in the dryrun. A scape of the entire log would take too long to count them all up manually.

🈯️ 
```
--stats
```
![image](https://user-images.githubusercontent.com/7077026/38517767-e8b0d7ea-3c32-11e8-966e-b2c7f256d24d.png)

And we can scrape total transfer file size 🤣

Fourdee commented:
### Tests:

no changes:
```
Total transferred file size: 21,110 bytes
```

---

5000 blank files
```
for ((i=0; i<5000; i++))
do

    echo '' > test.$i

done
```

```
Total transferred file size: 26,110 bytes
Number of regular files transferred: 5,001
```

DU MB result
```
root@DietPi:~# du -ms /root
20      /root
```

🈯️ 5001 * 4096 = 20484096 / 1024 / 1024 = 19.53515625 MB

So in theory our security measure could be ```Number of regular files transferred: 5,001``` * 4K

MichaIng commented:
@Fourdee 
About `Total transferred file size` the problem is, that it does not say anything about which files are replaced or possibly deleted, thus says nothing about the final disk space needed. It could be something between `Total transferred file size` and `current backup size + Total transferred file size + Number of regular files transferred * 4096`. But indeed this is the upper limit and therefore better than security factor guessing 👍 😃!

I will test and do PR, w8 😄.

Fourdee commented:
Test 2
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
total_transfered_file_size 26734
total_transfered_file_count 5001
total_4k_per_file_size 19 #MB

root@DietPi:~# dd if=/dev/zero of=/root/test bs=1M count=10
10+0 records in
10+0 records out
10485760 bytes (10 MB, 10 MiB) copied, 0.0424951 s, 247 MB/s

root@DietPi:~# dietpi-backup 1
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
total_transfered_file_size 10512494
total_transfered_file_count 5002
total_4k_per_file_size 19 #MB
```

---

mmm, rsync is really off with its total, with no file changes:
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
old_backup_size 594892
new_backup_size 560834
total_transfered_file_size 21
total_transfered_file_count 1
total_4k_per_file_size 4
end_result -33
```

Maybe we should just calculate the addition only, using ```total_transfered_file_size ```
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
total_transfered_file_size 20
total_transfered_file_count 1
total_4k_per_file_size 4
end_result 0
```

Fourdee commented:
@MichaIng 

> I will test and do PR, w8

Sorry, didn't see that. Lets both try it and compare at the end? :)

MichaIng commented:
Jep, as we both seem to have started already 😆.

MichaIng commented:
```
rsync --dry-run --stats $RSYNC_RUN_OPTIONS_SYSTEM_BACKUP "$FP_SOURCE_SYSTEM" "$FP_TARGET_BACKUP"/dietpi-backup_system/ > /tmp/dietpi-rsync-dry-run
local old_backup_size=$(du -bs "$FP_TARGET_BACKUP"/dietpi-backup_system | awk '{print $1}')
local new_backup_size=$(grep 'total size' /tmp/dietpi-rsync-dry-run | awk '{print $4}' | sed 's/,//g')
local target_fs_blocksize=$(stat -fc %s "$FP_TARGET_BACKUP"/dietpi-backup_system)
local transfered_file_count=$(grep 'Number of regular files transferred:' /tmp/dietpi-rsync-dry-run | awk '{print $6}' | sed 's/,//g')
local predicted_max_space_needed=$(( ($new_backup_size + $transfered_file_count * $target_fs_blocksize - $old_backup_size) / 1024 ** 2 ))
if G_CHECK_FREESPACE "$FP_TARGET_BACKUP"/dietpi-backup_system $predicted_max_space_needed; then
...
```
- Decided to use byte sizes until sum is done, to avoid rounding errors.
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_system | available=10335 MB | required=832 MB
[  OK  ] DietPi-Backup | rsync -aHv --delete --include-from=/tmp/.dietpi-backup_filter_inc_exc --log-file=/var/log/dietpi-backup.log / /mnt/dietpi-backup/dietpi-backup_system/
...
root@DietPi:~# du -sm /mnt/dietpi-backup/dietpi-backup_system/
816     /mnt/dietpi-backup/dietpi-backup_system/
```
- But w8, folders and links do also have block size as min. size, right? Hmm I get only the total number of "items" of the whole backup or the number of transferred regular files, but not the number of created folders and links 🤔.
**€: Ah: `Number of created files: 85 (reg: 52, dir: 32, special: 1)`**, this should cover what we need, we can calculate even more precise by counting `Number of created files - Number of deleted files`

Fourdee commented:
@MichaIng 

This is what I ended up with. 
- Only calculate additions in the total as rsync vs du results are too far off:
- Add 4K for each file in the transfer
```
				while true
				do

					# - system
					RSYNC_MODE_TEXT='Backup (system)'
					mkdir -p "$FP_TARGET_BACKUP"/dietpi-backup_system

					#	NB: working in K until end MB conversion, as, don't like using long long int, and, K should offer a good end result.
					rsync --dry-run --stats $RSYNC_RUN_OPTIONS_SYSTEM_BACKUP "$FP_SOURCE_SYSTEM" "$FP_TARGET_BACKUP"/dietpi-backup_system/ > /tmp/dietpi-backup_result
					local total_transfered_file_size=$(( $(grep -m1 'Total transferred file size' /tmp/dietpi-backup_result | sed 's/[^0-9]*//g') / 1024 ))
					local total_transfered_file_count=$(grep -m1 'Number of regular files transferred' /tmp/dietpi-backup_result | sed 's/[^0-9]*//g')
					local total_4k_per_file_size=$(( ( $total_transfered_file_count * 4096 ) / 1024 ))
					rm /tmp/dietpi-backup_result

					local end_result=$(( ( $total_transfered_file_size + $total_4k_per_file_size ) / 1024 ))

					#DEBUG:
					echo "total_transfered_file_size $total_transfered_file_size"
					echo "total_transfered_file_count $total_transfered_file_count"
					echo "total_4k_per_file_size $total_4k_per_file_size"
					echo "end_result $end_result"
					#DEBUG:

					if G_CHECK_FREESPACE "$FP_TARGET_BACKUP"/dietpi-backup_system $end_result; then
```

---

5000 file test:
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
total_transfered_file_size 4 K
total_transfered_file_count 5000
total_4k_per_file_size 20000 K
end_result 19 MB
```

MichaIng commented:
A bid rework:
```
rsync --dry-run --stats $RSYNC_RUN_OPTIONS_SYSTEM_BACKUP "$FP_SOURCE_SYSTEM" "$FP_TARGET_BACKUP"/dietpi-backup_system/ > /tmp/.dietpi-rsync-dry-run
local old_backup_size=$(du -bs "$FP_TARGET_BACKUP"/dietpi-backup_system | awk '{print $1}')
local new_backup_size=$(grep 'total size' /tmp/.dietpi-rsync-dry-run | awk '{print $4}' | sed 's/,//g')
local target_fs_blocksize=$(stat -fc %s "$FP_TARGET_BACKUP"/dietpi-backup_system)
local file_count=$(grep 'Number of files:' /tmp/.dietpi-rsync-dry-run | awk '{print $4}' | sed 's/,//g')
local max_space_needed=$(( ($new_backup_size + $file_count * $target_fs_blocksize - $old_backup_size) / 1024 ** 2 ))
```
- And moved `-v` (verbose) into logging options as well, as it's not needed for our calculations.
- Total file count is needed, as also total file size is used. Thought about using transferred file size and count, but again we can't use it as we don't know how large e.g. the deleted files...
```
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_system | available=10335 MB | required=852 MB
[  OK  ] DietPi-Backup | rsync -aH --delete --include-from=/tmp/.dietpi-backup_filter_inc_exc -v --log-file=/var/log/dietpi-backup.log / /mnt/dietpi-backup/dietpi-backup_system/
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_userdata | available=9518 MB | required=215 MB
[  OK  ] DietPi-Backup | rsync -aH --delete -v --log-file=/var/log/dietpi-backup.log /mnt/dietpi_userdata/ /mnt/dietpi-backup/dietpi-backup_userdata/
...
root@DietPi:~# du -ms /mnt/dietpi-backup/dietpi-backup_system/
816     /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -ms /mnt/dietpi-backup/dietpi-backup_userdata/
196     /mnt/dietpi-backup/dietpi-backup_userdata/
```

€: Hehe, as said, about just using transferred file stats: This could lead to total wrong values, if many files were deleted as well. Thus I ended up with calculating the maximum total new backup size with total size + block size for every file and folder and link that is present afterwards. Also blocksize could be different than 4k (? also 4k seems to be default, never found something different on linux so far), but can be estimated by `stat -fc %s <path>`.

Fourdee commented:
@MichaIng 

> This could lead to total wrong values, if many files were deleted as well. 

Yep, good point, got side tracked with the kids lol. Suppose its a "max possible size" value.

---

Yours is looking good 👍 

You can use ```| sed 's/[^0-9]*//g'``` to remove all non-numeric items 

eg:
```
local file_count=$(grep 'Number of files:' /tmp/.dietpi-rsync-dry-run | awk '{print $4}' | sed 's/,//g')
```

becomes
```
local file_count=$(grep 'Number of files:' /tmp/.dietpi-rsync-dry-run | sed 's/[^0-9]*//g)
```

MichaIng commented:
@Fourdee 
Sadly my scrape lines look like this:
```
Number of files: 15,588 (reg: 13,408, dir: 2,180)
total size is 162,402,959  speedup is 391.92 (DRY RUN)
```
Thus need to use awk as well 🙁.

Ah but `Total file size: 162,402,959 bytes` is available additionally with --stats. There `sed 's/[^0-9]*//g'` will work 🙂.

Fourdee commented:
@MichaIng 

mmm, i added the total file count + 4K on those to total of rsync dryrun, still not right on a backup with no changes:
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
old_backup_size 595020 K
total_file_count on new backup 12007 * 4K
new_backup_size 608863 K
```

^^ and thats not including the directory count + 4k

MichaIng commented:
@Fourdee 
Hmm, what you mean by not right? Of course we still cannot know the exact final size, as we don't know how much of the last 4k block is used by each theoretical file size. But at least it is a better worst case value, if e.g. 1 file is 49180673 bytes (1 byte filling a new 4k block) and all the others are empty.

Or did you define those empty files as before? Hmm in this case it should fit... will run a test as well.

Fourdee commented:
@MichaIng 

> Of course we still cannot know the exact final size, as we don't know how much of the last 4k block is used by each theoretical file size. 

Yep, i'am trying to achieve the impossible I think with perfect size match lol

Below seems to work well, although, 13MB+ offset on a backup with no changes.

```
				while true
				do

					# - system
					RSYNC_MODE_TEXT='Backup (system)'
					mkdir -p "$FP_TARGET_BACKUP"/dietpi-backup_system

					#	NB: working in K until end MB conversion, as, don't like using long long int, and, K should offer a good end result.

					local old_backup_size=$(du -ks "$FP_TARGET_BACKUP"/dietpi-backup_system | awk '{print $1}')

					rsync --dry-run --stats $RSYNC_RUN_OPTIONS_SYSTEM_BACKUP "$FP_SOURCE_SYSTEM" "$FP_TARGET_BACKUP"/dietpi-backup_system/ > /tmp/dietpi-backup_result
					local new_backup_size=$(( $(grep 'Total file size' /tmp/dietpi-backup_result | sed 's/[^0-9]*//g') / 1024 ))
					local total_file_count=$(( $(grep 'Number of files' /tmp/dietpi-backup_result | awk '{print $6}' | sed 's/[^0-9]*//g') ))
					new_backup_size=$(( $new_backup_size + ( ( $total_file_count * 4096 ) / 1024 ) ))

					local end_result=$(( ( $new_backup_size - $old_backup_size ) / 1024 ))
					rm /tmp/dietpi-backup_result

					#DEBUG:
					echo "old_backup_size $old_backup_size K"
					echo "total_file_count $total_file_count"
					echo "new_backup_size $new_backup_size K"
					echo "end_result $end_result M"
					#DEBUG:

					if G_CHECK_FREESPACE "$FP_TARGET_BACKUP"/dietpi-backup_system $end_result; then
```

MichaIng commented:
@Fourdee 
Jep, basically the same I have. Have even >70M offset on system and >50M offset on userdata backup without change, but it's closer than 1.2 factor at least 😉 and offset cannot be too small.
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
old_backup_size 816306972
new_backup_size 808763106
target_fs_blocksize 4096
file_count 20694
max_space_needed 73
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_system | available=9519 MB | required=73 MB
...
root@DietPi:~# du -bs /mnt/dietpi-backup/dietpi-backup_system/
816310686       /mnt/dietpi-backup/dietpi-backup_system/
```
Interesting, I created 10000 empty files and backed those up:
```
root@DietPi:~# du -bs /mnt/dietpi-backup/dietpi-backup_system/test/
Display all 10000 possibilities? (y or n)
176128  /mnt/dietpi-backup/dietpi-backup_system/test/
root@DietPi:~# G_TREESIZE /mnt/dietpi-backup/dietpi-backup_system/
176.0 KB /mnt/dietpi-backup/dietpi-backup_system/
172.0 KB /mnt/dietpi-backup/dietpi-backup_system/test
```
So it's definitely 10000 files inside, the blocksize is definitely 4k, as can be seen on `test` folder size (176k-172k=4k), but the overall folder size is way smaller then predicted 10000*4096=41M. Perhaps somehow several files are merged into one block in this case? 🤔

Fourdee commented:
@MichaIng 

Ok, i'll add directory count * 4K to code (as they do use up 4K of space), then send commit.
Need to get 6.7 out the door.

We can always improve it in 6.8 if required.

MichaIng commented:
@Fourdee 
As shown above, `Number of files` contains all folders, links and "special files" as well, thus not needed to add something, or is your output different?
`Number of files: 15,588 (reg: 13,408, dir: 2,180)` => 13,408 files + 2,180 folders
____________________________
Found something else confusing:
```
root@DietPi:~# du -bs /mnt/dietpi-backup/dietpi-backup_system/
816310686       /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -ks /mnt/dietpi-backup/dietpi-backup_system/
835108  /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -ms /mnt/dietpi-backup/dietpi-backup_system/
816     /mnt/dietpi-backup/dietpi-backup_system/
```
816 MiB = 835108 KiB okay, but how does 816310686 byte fit inside there?!?? Does `du` in the end calculate it's values wrong, mixing up KB/MB KiB/MiB? It it the same on your system? I will check on Buster 🤣. But in the end it should not make a too big difference... Anyway it is a warning based on a rough worst case calculation that still leaves decision to end user.
_____________________________
Btw. why hurry with 6.7? 6.6 is just a week ago and for my impression we could invest some pure testing and bugfix time for releases to avoid hotfix updates 😉.

Fourdee commented:
@MichaIng 

> As shown above, Number of files contain all folders, links and "special files" as well, thus not needed to add something, or is your output different?

Yep, i went with reg + dir, ignoring total and symlink.

> root@DietPi:~# du -bs /mnt/dietpi-backup/dietpi-backup_system/
816310686       /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -ks /mnt/dietpi-backup/dietpi-backup_system/
835108  /mnt/dietpi-backup/dietpi-backup_system/

Looks broken, good spot 👍 

http://whatsabyte.com/P1/byteconverter.htm

MichaIng commented:
@Fourdee 
Found it:
```
      -b, --bytes           equivalent to '--apparent-size --block-size=1'
      --apparent-size   print apparent sizes, rather than disk usage; although
                          the apparent size is usually smaller, it may be
                          larger due to holes in ('sparse') files, internal
                          fragmentation, indirect blocks, and the like
```
`du -sk / -sm` show disk usage, which is what we want, why ever `-bs` does not... Solution:
```
root@DietPi:~# du -s --block-size=1 /mnt/dietpi-backup/dietpi-backup_system/
855150592       /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -sk /mnt/dietpi-backup/dietpi-backup_system/
835108  /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -sm /mnt/dietpi-backup/dietpi-backup_system/
816     /mnt/dietpi-backup/dietpi-backup_system/
```
- With this my offset should be smaller like yours.

**€: Better at least 😃:**
```
[ INFO ] DietPi-Backup | Backing up to: /mnt/dietpi-backup
old_backup_size 855261184
new_backup_size 808862477
target_fs_blocksize 4096
file_count 20699
max_space_needed 36
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_system | available=9323 MB | required=36 MB
[  OK  ] DietPi-Backup | rsync -aH --delete --include-from=/tmp/.dietpi-backup_filter_inc_exc -v --log-file=/var/log/dietpi-backup.log / /mnt/dietpi-backup/dietpi-backup_system/
old_backup_size 205500416
new_backup_size 162402959
target_fs_blocksize 4096
file_count 15588
max_space_needed 19
[  OK  ] DietPi-Backup | Free space check: path=/mnt/dietpi-backup/dietpi-backup_userdata | available=9323 MB | required=19 MB
...
root@DietPi:~# du -s --block-size=1 /mnt/dietpi-backup/dietpi-backup_system/
855265280       /mnt/dietpi-backup/dietpi-backup_system/
root@DietPi:~# du -s --block-size=1 /mnt/dietpi-backup/dietpi-backup_userdata/
205500416       /mnt/dietpi-backup/dietpi-backup_userdata/
```

€2: But the mystery with the non 4k empty files is not solved by this 🤣:
```
root@DietPi:~# du -s --block-size=1 /mnt/test/
Display all 10000 possibilities? (y or n)
180224  /mnt/test/
root@DietPi:~# du -ks /mnt/test/
176     /mnt/test/
root@DietPi:~# du -bs /mnt/test/
176128  /mnt/test/
```

Fourdee commented:
> Btw. why hurry with 6.7? 6.6 is just a week ago and for my impression we could invest some pure testing and bugfix time for releases to avoid hotfix updates 

Good point, i'll move it to Saturday 👍

MichaIng commented:
@Fourdee 
Added another commit: https://github.com/Fourdee/DietPi/commit/0e89f375820cb326b8b383edde61dd4ca1eb4824
- As mentioned, we need to stay consistent with KiB/MiB (1024 / 1024 ** 2). I added this to comments as well to clarify.
- Moved verbose mode into logging options, as not needed and slowing down scraping for space estimation.
- Minor: `grep -m1`, formula transformation
________________________
- Checked everything again. Estimation works as expected and calculation results can be manually archived as well.
- But there is still ja little discrepancy, I am not sure about where it comes from:
```
root@DietPi:~# du -s --block-size=1 --apparent-size /mnt/dietpi-backup/dietpi-backup_userdata/
171356815       /mnt/dietpi-backup/dietpi-backup_userdata/
root@DietPi:~# du -s --block-size=1 /mnt/dietpi-backup/dietpi-backup_userdata/
205500416       /mnt/dietpi-backup/dietpi-backup_userdata/
root@DietPi:~# cat /tmp/dietpi-backup_result
...
Total file size: 162,402,959 bytes
...
total size is 162,402,959  speedup is 391.89 (DRY RUN)
```
- As far as I understood, `--apparent-size` gives the actual file bytes, not taking into account bock size. But for both, system and userdata backup, the bytes are still more than what is given as rsync result.
- Only idea: `although the apparent size is usually smaller, it may be larger due to holes in ('sparse') files, internal fragmentation, indirect blocks, and the like` But this should not be the case on nearly empty ext4 fs.
